File: src/main/java/com/zippeykeys/praisebe/util/ClassUtil.java
package com.zippeykeys.praisebe.util;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;

import org.apache.commons.lang3.ArrayUtils;
import org.jetbrains.annotations.Nullable;

import lombok.experimental.UtilityClass;
import lombok.var;


@UtilityClass
public class ClassUtil{
    @SuppressWarnings("unchecked")
    public static @Nullable <T> T newInstance(String className){
        try{
            return newInstance((Class<T>) Class.forName(className));
        }catch(ClassNotFoundException ignored){
            return null;
        }
    }

    public static @Nullable <T> T newInstance(Class<T> clazz){
        try{
            return clazz.newInstance();
        }catch(InstantiationException | IllegalAccessException ignored){
            return null;
        }
    }

    public static @Nullable <T> T newInstance(Constructor<T> constructor, Object... parameters){
        try{
            return constructor.newInstance(parameters);
        }catch(InstantiationException | IllegalAccessException | IllegalArgumentException
          | InvocationTargetException ignored){
            return null;
        }
    }

    public static @Nullable <T> Constructor<T> getConstructor(Class<T> clazz, Class<?>... parameterTypes){
        try{
            return clazz.getConstructor(parameterTypes);
        }catch(IllegalArgumentException | NoSuchMethodException | SecurityException ignored){
            return null;
        }
    }

    public static @Nullable Object getFieldValue(Field field){
        return getFieldValue(field, null);
    }

    public static @Nullable Object getFieldValue(Field field, @Nullable Object instance){
        try{
            return field.get(instance);
        }catch(IllegalArgumentException | IllegalAccessException ignored){
            return null;
        }
    }

    public static @Nullable Object callDeclaredMethod(Class<?> clazz, String methodName, Object instance){
        try{
            return clazz.getDeclaredMethod(methodName)
              .invoke(instance);
        }catch(IllegalAccessException | IllegalArgumentException | InvocationTargetException | NoSuchMethodException
          | SecurityException ignored){
            return null;
        }
    }

    public static Field[] getDeclaredFields(Class<?>... classes){
        var results = new Field[0];
        for(var clazz : classes){
            results = ArrayUtils.addAll(results, clazz.getDeclaredFields());
        }
        return results;
    }

    public static @Nullable Object getFieldValueByClass(Object caller, Class<?> gotten){
        try{
            return caller.getClass()
              .getMethod("get" + gotten.getSimpleName())
              .invoke(null);
        }catch(NoSuchMethodException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
          | SecurityException ignored){
            return null;
        }
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/util/ClassUtil.kt
package com.zippeykeys.praisebe.util

import org.apache.commons.lang3.ArrayUtils
import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.InvocationTargetException


@Suppress("UNCHECKED_CAST")
fun <T> newInstance(className: String): T? {
    return try {
        newInstance(Class.forName(className) as Class<T>)
    } catch (ignored: ClassNotFoundException) {
        null
    }

}

fun <T> newInstance(clazz: Class<T>): T? {
    return try {
        clazz.newInstance()
    } catch (ignored: InstantiationException) {
        null
    } catch (ignored: IllegalAccessException) {
        null
    }

}

fun <T> newInstance(constructor: Constructor<T>, vararg parameters: Any): T? {
    return try {
        constructor.newInstance(*parameters)
    } catch (ignored: InstantiationException) {
        null
    } catch (ignored: IllegalAccessException) {
        null
    } catch (ignored: IllegalArgumentException) {
        null
    } catch (ignored: InvocationTargetException) {
        null
    }

}

inline fun <reified T> getConstructor(vararg parameterTypes: Class<*>): Constructor<T>? {
    return try {
        T::class.java.getConstructor(*parameterTypes)
    } catch (ignored: IllegalArgumentException) {
        null
    } catch (ignored: NoSuchMethodException) {
        null
    } catch (ignored: SecurityException) {
        null
    }

}

@JvmOverloads
fun getFieldValue(field: Field, instance: Any? = null): Any? {
    return try {
        field.get(instance)
    } catch (ignored: IllegalArgumentException) {
        null
    } catch (ignored: IllegalAccessException) {
        null
    }

}

inline fun <reified T> callDeclaredMethod(methodName: String, instance: Any): Any? {
    return try {
        T::class.java.getDeclaredMethod(methodName)
                .invoke(instance)
    } catch (ignored: IllegalAccessException) {
        null
    } catch (ignored: IllegalArgumentException) {
        null
    } catch (ignored: InvocationTargetException) {
        null
    } catch (ignored: NoSuchMethodException) {
        null
    } catch (ignored: SecurityException) {
        null
    }

}

fun getDeclaredFields(vararg classes: Class<*>): Array<Field?> {
    var results = arrayOfNulls<Field>(0)
    for (clazz in classes) {
        results = ArrayUtils.addAll<Field>(results, *clazz.declaredFields)
    }
    return results
}

fun getFieldValueByClass(caller: Any, gotten: Class<*>): Any? {
    return try {
        caller.javaClass
                .getMethod("get" + gotten.simpleName)
                .invoke(null)
    } catch (ignored: NoSuchMethodException) {
        null
    } catch (ignored: IllegalAccessException) {
        null
    } catch (ignored: IllegalArgumentException) {
        null
    } catch (ignored: InvocationTargetException) {
        null
    } catch (ignored: SecurityException) {
        null
    }

}